File: 1 - Algorithmic Toolbox/Week_2/7 - Last Digit of the Sum of Fibonacci Numbers Again/last_digit_sum_fibonacci_numbers_again.py

def sum_of_fibonacci( fibonacci_index: int ) -> int:
    fibonacci_index %= 60
    
    if ( fibonacci_index <= 1 ):
        return fibonacci_index

    total: int = 1
    prev_val: int = 0
    curr_val = 1
    temp_val: int = 0

    for i in range( 1, fibonacci_index ):
        temp_val = prev_val
        prev_val = curr_val
        curr_val = ( temp_val + prev_val ) % 10
        total += curr_val
    return total % 10


def main( ) -> None:
    start, end = map( int, input( ).split( ) )
    a = sum_of_fibonacci( end % 60)
    b = sum_of_fibonacci( ( start - 1 ) % 60 )

    if ( a >= b ):
        print( a - b )
    else:
        print( 10 - ( b - a ) )

if __name__ == "__main__":
    main( )

File: 1 - Algorithmic Toolbox/2 - Algorithmic Warmup/7 - Last Digit of the Sum of Fibonacci Numbers Again/last_digit_sum_fibonacci_numbers_again.c
#include <stdio.h>
#include <stdlib.h>

long sum_of_fibonacci(long fibonacci_index)
{
    fibonacci_index %= 60;

    if (fibonacci_index == 0) {
        return 0;
    } else if (fibonacci_index == 1) {
        return 1;
    }

    long prev_val = 0;
    long curr_val = 1;
    long temp_val;
    long total = 1; // F(1) is already 1

    for (long i = 2; i <= fibonacci_index; i++) {
        temp_val = prev_val;
        prev_val = curr_val;
        curr_val = (temp_val + prev_val) % 10;
        total = (total + curr_val) % 10;
    }

    return total;
}

int main(void)
{
    long start;
    long end;
    scanf("%ld %ld", &start, &end);

    long sum_end = sum_of_fibonacci(end);
    long sum_start_minus_one = (start == 0) ? 0 : sum_of_fibonacci(start - 1);

    long result = (sum_end - sum_start_minus_one);
    if (result < 0) {
        result += 10;
    }

    printf("%ld\n", result);

    return 0;
}
